File: StriverA2Z/C++/2_SortingTechniques/insertion_sort.cpp
// O(n^2) - worst, avg
// O(n) - best

#include<bits/stdc++.h>

using namespace std;

void insertionSort(int arr[], int length) {
    for (int i = 0; i <= length - 1; i++) {
        int j = i;
        while (j > 0 && arr[j - 1] > arr[j]) {
            int temp = arr[j - 1];
            arr[j - 1] = arr[j];
            arr[j] = temp;
            j--;
        }
    }
}

int main() {
    int n;
    cout << "Enter the size of array:";
    cin >> n;

    int arr[n];
    for (int i = 0; i < n; i++) cin >> arr[i];

    insertionSort(arr, n);

    for (int x: arr) cout << x << " ";

    return 0;
}

File: StriverA2Z/Java/SortingAlgorithms/InsertionSort.java
public class InsertionSort {
    public static void main(String[] args) {
        int[] arr = {14, 9, 15, 12, 8, 6, 13};
        insertionSort(arr);
    }

    static void insertionSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int j = i;
            while (j > 0 && arr[j - 1] > arr[j]) {
                int temp = arr[j];
                arr[j] = arr[j - 1];
                arr[j - 1] = temp;
                j--;
            }
        }

        // for printing
        for (int x : arr) {
            System.out.print(x + " ");
        }
    }
}
